### DISCUSSION TITLE
Data Structure: Transactions and Holdings

### DISCUSSION THREAD

**Discussion by mkitzmann** [GoPlan-Finance/goplan-app#35] [Q&A]
Every user would have a number of transactions which would results in that users holdings. Do we want to save only the transactions and calculate the holdings or do we want to regenerate the holdings each time a transaction is made?
With possibly thousands of transactions it would be very ineffiecent to calculate the holding every time, but with generating them at a fixed time, we do not have a single source of truth, which could eventually lead to an inconsistent state.

**Comment by sadortun** [comment]
You are right, and it may be a bit more complicated than that.

Few things to consider: 
- `Holding` amounts and $$ and  `Account` values and $$ are only accessible from the user since they are encrypted client-side
- There is a few levels of generalization : `Execution` ->  `Order` -> `Holding` -> `Account` / `WatchList` -> `Global Totals` 
- Each of theses generalizations requires both to sum up the childs, and allow a direct entry. Example:
   -  I have 1000$ in my account right now, and i hold  10 `TSLA`
   - I bougth  5 more `TSLA` yesturday at `600$`
   - I bougth  3 more `TSLA` today at `550$`
- To update the `Account` when a `Holding` is modified we can simply  filter/apply when `Holding.updatedAt > Account.updatedAt`
- When a user delete a `Holding`, we probably need to only set a flag to `false` or not allow


I did had a similar thing on a different project and my solution was to handle everything via [Time Series](https://www.mongodb.com/blog/post/time-series-data-and-mongodb-part-2-schema-design-best-practices)

So at the end, you end up with two collections: 
- TimeSeries which contains the totals for various durations (`year`, `month`, `day` and maybe `hour`). FYI Each duration contains data from the smaller time scale. For example, the `year` Series contains 12 months
- TimeSeriesData  which contains the source data for each series. This is useful in the case we need to rebuild the series. We dont have to store everything in one single `TimeSeriesData` but we need to keep the reference data in an `Order` or `Holding` collection for example

After that, its fairly easy to compare the  `TimeSeriesData.updatedAt > TimeSerie.updatedAt` and update the TimeSeries.
